env/src/main/java/com/sap/cloud/security/config/Environments.java

```diff
@@ -6,20 +6,29 @@
 package com.sap.cloud.security.config;
 
 import com.sap.cloud.security.config.cf.CFEnvironment;
+import com.sap.cloud.security.token.ProviderNotFoundException;
+import org.slf4j.LoggerFactory;
 
 import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Scanner;
+import java.util.ServiceLoader;
 
 /**
  * Central entry point to access the current SAP Cloud Platform
  * {@link Environment}.
  */
 public class Environments {
-
-	private static final Environment cfEnvironment = CFEnvironment.getInstance(); // singleton
+	static List<EnvironmentLoader> environments = new ArrayList() {
+		{
+			ServiceLoader.load(EnvironmentLoader.class).forEach(this::add);
```

Comment:
Not clear how clients can overrule the default loader as all matching loaders will be added to environments, but only the one with index 0 is taken as current.
What about replacing the ServiceLoader API and just provide a static setter which writes a single reference (default to DefaultEnvironmentLoader) instead of a list? This also helps in a Spring context in which ServiceLoader implementations are hard to combine with initialization of Spring beans.

Reply:
Hi @BraunMatthias  thanks for the feedback! We will revisit your first question soon. 

What do you mean with `This also helps in a Spring context in which ServiceLoader implementations are hard to combine with initialization of Spring beans`?
In our AMS client lib, we heavily combine Service loader and spring-beans....

Reply:
My specific EnvironmentLoader needs some of our Spring beans as prerequisite (as input parameter) which also puts some constraints on creation order etc. Ideally, the EnvironmentLoader  should be bean itself that dependes on other beans in my setup. On the other hand, it seems there is no control about service loader object creation and input parameters (the creation is triggered when the Environments class is loaded, no input parameters).
Can you point me to an example how to deal with objects loaded via ServiceLoader that rely on Spring bean capabilities?